src/Features/Core/Portable/MetadataAsSource/FileSystemMetadataDocumentPersister.cs

```diff
@@ -0,0 +1,177 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.MetadataAsSource;
+
+internal class FileSystemMetadataDocumentPersister : IMetadataDocumentPersister
```

Comment:
most of the implementation here is the existing file based support extracted out

Reply:
sealed?